### PR TITLE
MBS-7725 Switch to graphite in teamcity metrics [MERGE]

### PR DESCRIPTION
**Pull request by eugene-krivobokov**
It allows to override event timestamp.

### CODE REVIEW

**Review by dsvoronin — APPROVED** [review]

File: subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/Graphite.kt [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ * https://graphite.readthedocs.io/en/latest/terminology.html#term-value
+ */
+fun graphiteSeriesElement(value: String) = value.replace(invalidSymbols, "_")
+
+/**
+ * https://graphite.readthedocs.io/en/latest/terminology.html#term-series
+ */
+fun graphiteSeries(vararg element: String) = element.joinToString(
```
Comment (sergioboi) [inline subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/Graphite.kt]: Advice: add validation for empty element

File: subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/Graphite.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package com.avito.android.graphite
+
+/**
+ * https://graphite.readthedocs.io/en/latest/terminology.html#term-value
+ */
+fun graphiteSeriesElement(value: String) = value.replace(invalidSymbols, "_")
```
Comment (sergioboi) [inline subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/Graphite.kt]: internal?

File: subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/GraphiteConfig.kt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package com.avito.android.graphite
+
+import java.io.Serializable
+
+data class GraphiteConfig(
+    val isEnabled: Boolean,
```
Comment (sergioboi) [inline subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/GraphiteConfig.kt:9]: Not intuitive parameter. Need comment

File: subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/GraphiteSender.kt [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+interface GraphiteSender {
+
+    /**
+     * Incubating
+     */
+    fun send(metric: GraphiteMetric)
+
+    class Impl(
+        private val config: GraphiteConfig,
+        private val logger: (String, Throwable?) -> Unit
```
Comment (sergioboi) [inline subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/GraphiteSender.kt]: It's look like tech debt. Let's use normal logger

File: subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/GraphiteSender.kt [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            send(metricName, metric.value, metric.timestamp)
+        }
+
+        private fun send(path: String, value: String, timestamp: Long) {
+            // https://graphite.readthedocs.io/en/latest/feeding-carbon.html#the-plaintext-protocol
+            val message = "$path $value $timestamp\n"
+
+            logger.invoke("graphite: $message", null)
+
+            if (!config.isEnabled) return
```
Comment (sergioboi) [inline subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/GraphiteSender.kt]: It's look ugly. I can see in log send info but send woun't be happen
```suggestion
             if (!config.isEnabled) {
              // ignore info
             } else {
             // send info
             }
```

File: subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/GraphiteSender.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+            try {
+                socket().use { socket ->
```
Comment (sergioboi) [inline subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/GraphiteSender.kt:29]: Advice: Highly tied. Hide transport details

File: subprojects/gradle/build-metrics/src/main/kotlin/com/avito/android/plugin/build_metrics/CollectTeamcityMetricsWorkerAction.kt [hunk 7]
```diff
@@ -16,21 +16,21 @@ abstract class CollectTeamcityMetricsWorkerAction : WorkAction<Parameters> {
...
         fun getLogger(): Property<CILogger>
     }
 
     override fun execute() {
         require(!parameters.getBuildId().orNull.isNullOrBlank()) { "teamcity buildId property must be set" }
         val logger = parameters.getLogger().get()
-        val statsd: StatsDSender = StatsDSender.Impl(parameters.getStatsdConfig().get()) { msg, _ ->
-            logger.info(msg)
+        val graphite: GraphiteSender = GraphiteSender.Impl(parameters.getGraphiteConfig().get()) { msg, error ->
+            logger.info(msg, error)
```
Comment (sergioboi) [inline subprojects/gradle/build-metrics/src/main/kotlin/com/avito/android/plugin/build_metrics/CollectTeamcityMetricsWorkerAction.kt]: Why do we need this info?

File: subprojects/gradle/graphite-config/src/main/kotlin/com/avito/android/stats/ProjectExtensions.kt [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+val Project.graphite: Provider<GraphiteSender> by ProjectProperty.lazy(scope = ROOT_PROJECT) { project ->
+    val logger = project.ciLogger
+
+    Providers.of<GraphiteSender>(GraphiteSender.Impl(
+        config = config(project),
+        logger = { message, error -> if (error != null) logger.info(message, error) else logger.debug(message) }
+    ))
+}
+
+val Project.graphiteConfig: Provider<GraphiteConfig> by ProjectProperty.lazy(scope = ROOT_PROJECT) { project ->
```
Comment (sergioboi) [inline subprojects/gradle/graphite-config/src/main/kotlin/com/avito/android/stats/ProjectExtensions.kt]: internal or make properties keys const and public

**Review by sergioboi — CHANGES_REQUESTED** [review]

File: subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/Graphite.kt (hunk 2 above)
Comment (eugene-krivobokov) [reply subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/Graphite.kt]: nope, it's intended to use in clients

File: subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/Graphite.kt (hunk 1 above)
Comment (eugene-krivobokov) [reply subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/Graphite.kt]: Removed these functions. It was a copy from statsd module

File: subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/Graphite.kt (hunk 2 above)
Comment (eugene-krivobokov) [reply subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/Graphite.kt]: Removed these functions. It was a copy from statsd module

File: subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/GraphiteConfig.kt (hunk 3 above)
Comment (eugene-krivobokov) [reply subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/GraphiteConfig.kt:9]: fixed

File: subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/GraphiteSender.kt (hunk 4 above)
Comment (eugene-krivobokov) [reply subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/GraphiteSender.kt]: fixed

File: subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/GraphiteSender.kt (hunk 5 above)
Comment (eugene-krivobokov) [reply subprojects/common/graphite/src/main/kotlin/com/avito/android/graphite/GraphiteSender.kt]: fixed. I've added "debug" mode in config

**Review by sergioboi — APPROVED** [review]

**Pull request merged**